File: src/Pcl.NET/PointCloudXYZRGBA.cs
namespace Pcl.NET
{

    public class PointCloudXYZRGBA : PointCloud<PointXYZRGBA>
    {
        private readonly VectorXYZRGBA _points;

        public override int Width
        {
            get
            {
                ThrowIfDisposed();
                return (int)Invoke.pointcloud_xyzrgba_get_width(_ptr);
            }
            set
            {
                ThrowIfDisposed();
                Invoke.pointcloud_xyzrgba_set_width(_ptr, (uint)value);
            }
        }
        public override int Height
        {
            get
            {
                ThrowIfDisposed();
                return (int)Invoke.pointcloud_xyzrgba_get_height(_ptr);
            }
            set
            {
                ThrowIfDisposed();
                Invoke.pointcloud_xyzrgba_set_height(_ptr, (uint)value);
            }
        }
        public override bool IsDense
        {
            get
            {
                ThrowIfDisposed();
                return Invoke.pointcloud_xyzrgba_get_is_dense(_ptr);
            }
            set
            {
                ThrowIfDisposed();
                Invoke.pointcloud_xyzrgba_set_is_dense(_ptr, value);
            }
        }
        public override Vector<PointXYZRGBA> Points => _points;
        public override long Count
        {
            get
            {
                ThrowIfDisposed();
                return (long)Invoke.pointcloud_xyzrgba_size(_ptr);
            }
        }
        public override bool IsOrganized
        {
            get
            {
                ThrowIfDisposed();
                return Invoke.pointcloud_xyzrgba_is_organized(_ptr);
            }
        }

        private PointCloudXYZRGBA(IntPtr ptr)
        {
            _ptr = ptr;
            _points = new VectorXYZRGBA(Invoke.pointcloud_xyzrgba_points(_ptr));
        }

        internal PointCloudXYZRGBA(IntPtr ptr, bool suppressDispose)
          : this(ptr)
        {
            _suppressDispose = suppressDispose;
        }

        public PointCloudXYZRGBA()
          : this(Invoke.pointcloud_xyzrgba_ctor())
        {
        }

        public PointCloudXYZRGBA(int width, int height)
          : this(Invoke.pointcloud_xyzrgba_ctor_wh((uint)width, (uint)height))
        {
        }

        public PointCloudXYZRGBA(PointCloudXYZRGBA other)
          : this(Invoke.pointcloud_xyzrgba_ctor_indices(other, IntPtr.Zero))
        {
        }

        public PointCloudXYZRGBA(PointCloudXYZRGBA other, VectorInt indices)
          : this(Invoke.pointcloud_xyzrgba_ctor_indices(other, indices))
        {
        }

        public unsafe override void Add(PointXYZRGBA value)
        {
            ThrowIfDisposed();
            Invoke.pointcloud_xyzrgba_add(_ptr, &value);
        }

        public unsafe override ref PointXYZRGBA At(int col, int row)
        {
            ThrowIfDisposed();
            ThrowHelper.ThrowUnorganizedPointCloudfCondition_CantUse2DIndexing(!IsOrganized);
            ThrowHelper.ThrowArgumentOutOfRangeIfCondition_IndexMustBeLessException(col >= Width, nameof(col));
            ThrowHelper.ThrowArgumentOutOfRangeIfCondition_IndexMustBeLessException(row >= Height, nameof(row));
            return ref System.Runtime.CompilerServices.Unsafe.AsRef<PointXYZRGBA>(Invoke.pointcloud_xyzi_at_colrow(_ptr, col, row));
        }

        public PointCloudXYZRGBA Downsample(int factor)
        {
            ThrowIfDisposed();
            PointCloudXYZRGBA output = new PointCloudXYZRGBA(this.Width, this.Height);
            Invoke.pointcloud_xyzrgba_downsample(_ptr, factor, output);
            return output;
        }

        public static PointCloudXYZRGBA operator +(PointCloudXYZRGBA a, PointCloudXYZRGBA b)
        {
            return Concatenate(a, b);
        }

        #region Static

        public static PointCloudXYZRGBA Concatenate(PointCloudXYZRGBA a, PointCloudXYZRGBA b)
        {
            a.ThrowIfDisposed();
            b.ThrowIfDisposed();

            PointCloudXYZRGBA outpc = new PointCloudXYZRGBA();

            Invoke.pointcloud_xyzrgba_concatenate(a, b, outpc);

            return outpc;
        }

        public static PointCloudXYZRGBA Load(string filename)
        {
            return IO.LoadPointCloudXYZRGBA(filename);
        }

        #endregion

        #region Dispose
        protected override void DisposeObject()
        {
            if (!_suppressDispose)
            {
                Invoke.pointcloud_xyzrgba_delete(ref _ptr);
            }
        }
        #endregion
    }
}

File: src/pcl-bindings/PointCloudXYZRGBA.cpp
#include "PointCloud_Common.h"
#include "export.h"
#include "pcl/pcl_base.h"
#include "pcl/point_types.h"

using namespace pcl;
using namespace std;

using point_t = PointXYZRGBA;
using pointcloud_t = PointCloud<point_t>;
using point_vector = vector<point_t, Eigen::aligned_allocator<point_t>>;

EXPORT(pointcloud_t*) pointcloud_xyzrgba_ctor()
{
    return new pointcloud_t();
}

EXPORT(pointcloud_t*) pointcloud_xyzrgba_ctor_wh(uint32_t width, uint32_t height)
{
    return new pointcloud_t(width, height);
}

EXPORT(pointcloud_t*) pointcloud_xyzrgba_ctor_indices(pointcloud_t* cloud, vector<int>* indices)
{
    if (indices == NULL)
        return new pointcloud_t(*cloud);
    else
        return new pointcloud_t(*cloud, *indices);
}

EXPORT(void) pointcloud_xyzrgba_delete(pointcloud_t** ptr)
{
    delete* ptr;
    *ptr = NULL;
}

EXPORT(point_t*) pointcloud_xyzrgba_at_colrow(pointcloud_t* ptr, int col, int row)
{
    return &(ptr->at(col, row));
}

EXPORT(void) pointcloud_xyzrgba_add(pointcloud_t* ptr, point_t* value)
{
    //the value needs to be aligned to be pushed into the cloud, so do that.
    point_t deref;
    memcpy(&deref, value, sizeof(point_t));
    ptr->push_back(deref);
}

EXPORT(size_t) pointcloud_xyzrgba_size(pointcloud_t* ptr)
{
    return ptr->size();
}

EXPORT(void) pointcloud_xyzrgba_clear(pointcloud_t* ptr)
{
    ptr->clear();
}

EXPORT(uint32_t) pointcloud_xyzrgba_get_width(pointcloud_t* ptr)
{
    return ptr->width;
}

EXPORT(void) pointcloud_xyzrgba_set_width(pointcloud_t* ptr, uint32_t width)
{
    ptr->width = width;
}

EXPORT(uint32_t) pointcloud_xyzrgba_get_height(pointcloud_t* ptr)
{
    return ptr->height;
}

EXPORT(void) pointcloud_xyzrgba_set_height(pointcloud_t* ptr, uint32_t height)
{
    ptr->height = height;
}

EXPORT(int32_t) pointcloud_xyzrgba_is_organized(pointcloud_t* ptr)
{
    return ptr->isOrganized();
}

EXPORT(point_vector*) pointcloud_xyzrgba_points(pointcloud_t* ptr)
{
    return &ptr->points;
}

EXPORT(point_t*) pointcloud_xyzrgba_data(pointcloud_t* ptr)
{
    return ptr->points.data();
}

EXPORT(void) pointcloud_xyzrgba_downsample(pointcloud_t* ptr, int factor, pointcloud_t* output)
{
    downsample<point_t>(ptr, factor, output);
}

EXPORT(void) pointcloud_xyzrgba_set_is_dense(pointcloud_t* ptr, int value)
{
    ptr->is_dense = value;
}

EXPORT(int) pointcloud_xyzrgba_get_is_dense(pointcloud_t* ptr)
{
    return ptr->is_dense;
}

EXPORT(void) pointcloud_xyzrgba_concatenate(pointcloud_t* ptr1, pointcloud_t* ptr2, pointcloud_t* ptr_out)
{
    pointcloud_t::concatenate(*ptr1, *ptr2, *ptr_out);
}
